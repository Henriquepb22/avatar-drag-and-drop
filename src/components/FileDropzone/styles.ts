import styled, { css } from 'styled-components'

type WrapperProps = {
    isSaved: boolean
}

export const Wrapper = styled.label<WrapperProps>`
    ${({ theme, isSaved }) => css`
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;
        position: absolute;
        left: 0;
        top: 0;
        right: 0;
        bottom: 0;
        font-size: ${theme.font.sizes.small};
        line-height: 180%;

        &:hover {
            cursor: pointer;
        }

        &:focus-within {
            outline: -webkit-focus-ring-color auto 1px;
        }

        ${isSaved &&
        css`
            padding-left: calc(${theme.sizes.image} + ${theme.spacings.large});
        `}
    `}
`

export const LabelText = styled.p`
    ${({ theme }) => css`
        color: ${theme.colors.grayDarkness};
        font-weight: ${theme.font.medium};

        > img {
            margin-right: ${theme.spacings.small};
        }
    `}
`

export const LabelPlaceholder = styled.p`
    ${({ theme }) => css`
        color: ${theme.colors.grayDark};
    `}
`

export const Input = styled.input`
    height: 0;
    width: 0;
    border: none;
    opacity: 0;
`
