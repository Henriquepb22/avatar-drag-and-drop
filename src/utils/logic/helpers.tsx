const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
