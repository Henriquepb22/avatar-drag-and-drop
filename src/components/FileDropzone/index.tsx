import React from 'react'

import paintingIcon from 'assets/img/icons/painting.svg'
import * as S from './styles'

type FileDropzoneProps = {
    handleFileUpload: (file: File) => void
    isSaved?: boolean
}

const FileDropzone = ({
    handleFileUpload,
    isSaved = false,
}: FileDropzoneProps) => {
    const onDrop = (e: React.DragEvent<HTMLLabelElement>) => {
        e.preventDefault()
        e.dataTransfer.files?.length &&
            handleFileUpload(e.dataTransfer.files[0])
    }

    const onDragOver = (e: React.DragEvent<HTMLLabelElement>) => {
        e.preventDefault()
    }

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault()
        e.currentTarget.files?.length &&
            handleFileUpload(e.currentTarget.files[0])
    }

    return (
        <S.Wrapper
            htmlFor="uploadInput"
            onDrop={onDrop}
            onDragOver={onDragOver}
            isSaved={isSaved}
        >
            <S.LabelText>
                <img src={paintingIcon} alt="mountain and sun" />
                Organization Logo
            </S.LabelText>
            <S.LabelPlaceholder>
                Drop the image here or click to browse.
            </S.LabelPlaceholder>
            <S.Input
                type="file"
                id="uploadInput"
                accept="image/*"
                onChange={onChange}
                tabIndex={0}
            />
        </S.Wrapper>
    )
}

export default FileDropzone
