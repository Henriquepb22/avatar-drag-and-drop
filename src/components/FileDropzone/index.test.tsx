import { screen, waitFor, fireEvent } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import userEvent from '@testing-library/user-event'

import FileDropzone from '.'

const imageFile = new File(['image file'], 'image.png', {
    type: 'image/png',
})

describe('<FileDropzone />', () => {
    it('should render file dropzone correctly', () => {
        const { container } = renderWithTheme(
            <FileDropzone handleFileUpload={jest.fn()} />
        )

        expect(
            screen.getByRole('img', { name: /mountain and sun/i })
        ).toBeInTheDocument()
        expect(screen.getByLabelText(/organization logo/i)).toBeInTheDocument()
        expect(
            screen.getByLabelText(/drop the image here/i)
        ).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should upload image file by clicking on input', async () => {
        const handleFileUpload = jest.fn()
        renderWithTheme(<FileDropzone handleFileUpload={handleFileUpload} />)

        const fileInput = screen.getByLabelText(
            /drop the image here/i
        ) as HTMLInputElement

        userEvent.upload(fileInput, imageFile)

        await waitFor(() => {
            expect(fileInput.files).toHaveLength(1)
            expect(fileInput.files?.item(0)).toStrictEqual(imageFile)
            expect(handleFileUpload).toHaveBeenCalledWith(imageFile)
        })
    })

    it('should upload image file by dragging and dropping it', async () => {
        const handleFileUpload = jest.fn()
        renderWithTheme(<FileDropzone handleFileUpload={handleFileUpload} />)

        const fileInput = screen.getByLabelText(/drop the image here/i)

        fireEvent.drop(fileInput, {
            dataTransfer: {
                files: [imageFile],
            },
        })

        await waitFor(() => {
            expect(handleFileUpload).toHaveBeenCalledWith(imageFile)
        })
    })

    it('should prevent default on dragover event', () => {
        renderWithTheme(<FileDropzone handleFileUpload={jest.fn()} />)

        const fileInput = screen.getByLabelText(/drop the image here/i)

        const isPrevented = fireEvent.dragOver(fileInput)
        expect(isPrevented).toBe(false)
    })

    it('should be accessible by tab', () => {
        renderWithTheme(<FileDropzone handleFileUpload={jest.fn()} />)

        const fileInput = screen.getByLabelText(/drop the image here/i)

        expect(fileInput).not.toHaveFocus()
        userEvent.tab()
        expect(fileInput).toHaveFocus()
    })
})
