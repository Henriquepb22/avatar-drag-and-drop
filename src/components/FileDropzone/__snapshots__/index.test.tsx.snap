// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileDropzone /> should render file dropzone correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  font-size: 1.4rem;
  line-height: 180%;
}

.c0:hover {
  cursor: pointer;
}

.c0:focus-within {
  outline: -webkit-focus-ring-color auto 1px;
}

.c1 {
  color: #495567;
  font-weight: 500;
}

.c1 > img {
  margin-right: 1.2rem;
}

.c2 {
  color: #677489;
}

.c3 {
  height: 0;
  width: 0;
  border: none;
  opacity: 0;
}

<label
  class="c0"
  for="uploadInput"
>
  <p
    class="c1"
  >
    <img
      alt="mountain and sun"
      src="painting.svg"
    />
    Organization Logo
  </p>
  <p
    class="c2"
  >
    Drop the image here or click to browse.
  </p>
  <input
    accept="image/*"
    class="c3"
    id="uploadInput"
    tabindex="0"
    type="file"
  />
</label>
`;
